pages/slice/language-reference/slice-grammar.md

```diff
@@ -2,14 +2,572 @@
+
+```ebnf {% showTitle=false %}
+TypeRef
+    : LocalAttribute* TypeRefDefinition "?"?
+    ;
+
+TypeRefDefinition
+    : Primitive | Sequence | Dictionary | RelativeIdentifier | GlobalIdentifier
+    ;
+```
+
+### Tags
+
+Tags consist of the `tag` keyword, followed by a positive integer wrapped in a pair of parenthesis.
+This integer must between `0` and `2,147,483,647` (the maximum value of a signed 32bit integer).
+Tags can only be applied to [fields][field] and [parameters][parameter], and only if the type of that field/parameter:
+- is optional (it ends with `?` symbol)
+- is not a class, and does not use classes internally
+
+```ebnf {% showTitle=false %}
+Tag
+    : "tag" "(" SignedInteger ")"
+    ;
+```
+
+### Identifiers
+
+Slice supports three forms of identifiers:
+- An unscoped identifier. These consist of a letter, followed by any number of letters, digits, or underscores.
+- A relatively scoped identifier. These consist of one or more unscoped identifiers, separated by `::` tokens.
+- A globally scoped identifier. These consist of a `::` token, followed by a relatively scoped identifier.
+
+```ebnf {% showTitle=false %}
+Identifier
+    : identifier
+    ;
+
+RelativeIdentifier
+    : identifier ("::" identifier)*
+    ;
+
+GlobalIdentifier
+    : ("::" identifier)+
+    ;
+```
+
+Examples of valid identifiers:
+
+```slice
+// Examples of unscoped identifiers
+foo9
+_bar_
+BAZ
+
+// Examples of relatively scoped identifiers
+foo9
+foo9::_bar_::BAZ
+
+// Examples of globally scoped identifiers
+::foo9
+::foo9::_bar_::BAZ
+```
+
+All Slice [keywords] satisfy the definition of an unscoped identifier.
+Hence the lexer always gives priority to producing keyword tokens, even when it could produce an identifier token.
+To use a keyword as an identifier, you can escape the keyword by prefixing it with a backslash.
+For scoped identifiers, each individual identifier must be escaped this way.
+It is legal to escape an identifier that is not a keyword, though doing so has no effect.
+
+```slice
+struct   // The struct keyword
+\struct  // An unscoped identifier
+\foo     // An unscoped identifier
+
+::\struct::\module // A globally scoped identifier
+```
+
+### Integer literals
+
+Slice supports three forms of integer literals:
+- A decimal literal for base-10 numbers. These consist of one or more digits between `0` and `9`.
+- A hexadecimal literal for base-16 numbers. These start with the prefix `0x`, followed by one or more digits between `0` and `9` or letters between `a` and `f`. Hexadecimal literals are not case-sensitive.
+- A binary literal for base-2 numbers. These start with the prefix `0b`, followed by one or more digits that are either `0` or `1`.
+
+Additionally, underscores can appear between any two characters of an integer literal.
+These underscores carry no semantic meaning and are discarded by the parser.
+
+Examples of valid integer literals:
```

Comment:
I think the opposite : v) The focus of this section is solely integer literals. Suddenly bringing `tag`s into it would just blur what's being defined here I worry.

Plus, it's still the same problem:
You can't just put an integer literal in the middle of your Slice file.
But you also can't just put a tag in the middle of your Slice file.